sdk/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+    fn default() -> Self {
+        Self {
+            batch_size: 1,
+            starting_cursor: None,
+            finality: None,
+            filter: Some(Filter::default()),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ApibaraIndexerError {
+    #[error("Failed to build indexer")]
+    FailedToBuildIndexer,
+    #[error(transparent)]
+    TonicError(#[from] tonic::transport::Error),
+    #[error(transparent)]
+    FailedToConfigureStream(#[from] SendError<Configuration>),
+    #[error(transparent)]
+    StreamError(#[from] tonic::Status),
+}
+
+pub struct ApibaraIndexer<StreamClient> {
+    client: Option<StreamClient>,
+    configuration: Option<Configuration>,
+    configuration_channel: Option<(Sender<Configuration>, Receiver<Configuration>)>,
+    inner_channel: Option<(Sender<StreamDataRequest>, Receiver<StreamDataRequest>)>,
+}
+
+impl<StreamClient> Default for ApibaraIndexer<StreamClient> {
+    fn default() -> Self {
+        Self {
+            client: None,
+            configuration: None,
+            configuration_channel: None,
+            inner_channel: None,
+        }
+    }
+}
+
+pub struct StreamHandler {
+    stream_id: u64,
+    configuration_rx: Receiver<Configuration>,
+    inner: StreamingClientWrapper,
+    inner_tx: Sender<StreamDataRequest>,
+}
+
+#[derive(Debug)]
+#[pin_project]
+struct StreamingClientWrapper {
+    #[pin]
+    inner: Streaming<StreamDataResponse>,
+}
+
+impl ApibaraIndexer<StreamClient<Channel>> {
+    pub fn with_client(mut self, client: StreamClient<Channel>) -> Self {
+        self.client = Some(client);
+        self
+    }
+
+    pub fn with_configuration(mut self, configuration: Configuration) -> Self {
+        self.configuration = Some(configuration);
+        self
+    }
+    pub fn with_configuration_channel(
+        mut self,
+        configuration_channel: (Sender<Configuration>, Receiver<Configuration>),
+    ) -> Self {
+        self.configuration_channel = Some(configuration_channel);
+        self
+    }
+
+    pub fn with_inner_channel(
+        mut self,
+        inner_channel: (Sender<StreamDataRequest>, Receiver<StreamDataRequest>),
+    ) -> Self {
+        self.inner_channel = Some(inner_channel);
+        self
+    }
+
+    /// Apibara indexer SDK
+    ///
+    /// let base_configuration = Configuration::default();
+    /// let (mut stream, _configuration_handle) =
+    ///     ApibaraIndexer::default().build().await?;
+    /// while let Some(response) = stream.next().await {
+    ///     // Handle your own business logic in there.
+    /// }
+    pub async fn build(
+        self,
+        indexer_url: Option<String>,
+    ) -> Result<(StreamHandler, Sender<Configuration>), ApibaraIndexerError> {
+        let mut client = self.client.unwrap_or(
+            StreamClient::connect(indexer_url.unwrap_or("http://0.0.0.0:7171".into())).await?,
+        );
+        let configuration = self.configuration.unwrap_or(Configuration::default());
```

Comment:
In my opinion, if the user didn't specify the starting configuration, then it shouldn't send any.

The code will be  something like:

```rs
if let Some(configuration) = self.configuration) {
  configuration_tx.send(configuration).await?;
}
```